### PR TITLE
feat: unpin redis

### LINKED ISSUE #2208: Bug: Fix issues with py-redis >=4.5.0
### Description

We currently pin redis to version `4.4.4` because any version above that causes test failues. I was not able to determine the reason for this test failures, but this will cause issues for users who wish to use the newer versions of the library (version 5.0.0 is already out). 

Tasks:

1. unpin the redis version.
2. fix the issues causing the test failures.

### URL to code causing the issue

_No response_

### MCVE

_No response_

### Steps to reproduce

_No response_

### Screenshots

_No response_

### Logs

_No response_

### Litestar Version

v2.0.0

### Platform

- [X] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---

## Funding
* If you would like to see an issue prioritized, make a pledge towards it!
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/litestar-org/litestar/issues/2208">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2208/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by euri10**
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

- unpin redis

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

- Fixes #2208

### CODE REVIEW

**Comment by euri10** [comment]
for my ref, failing tests after unpin

tests.unit.test_channels.test_plugin.test_pub_sub_wait_published
tests.unit.test_channels.test_plugin.test_pub_sub_non_blocking
tests.unit.test_channels.test_plugin.test_pub_sub_run_in_background
tests.unit.test_channels.test_plugin.test_create_ws_route_handlers
tests.unit.test_channels.test_plugin.test_create_ws_route_handlers_arbitrary_channels_allowed

**Comment by Goldziher** [comment]
> the issue first arises in 4.5.0 and only on the tests above for the `redis_pub_sub_backend`, `memory` and `stream_backend` work fine
> 
> from https://github.com/redis/redis-py/compare/v4.4.4...v4.5.0 and given the logs of the failing tests :
> 
> ```
> FAILED [100%]
> tests/unit/test_channels/test_plugin.py:66 (test_pub_sub_wait_published[redis:pubsub])
> channels_backend = <litestar.channels.backends.redis.RedisChannelsPubSubBackend object at 0x7fd7822b9350>
> 
>     async def test_pub_sub_wait_published(channels_backend: ChannelsBackend) -> None:
> >       async with ChannelsPlugin(backend=channels_backend, channels=["something"]) as plugin:
> 
> test_channels/test_plugin.py:68: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../../litestar/channels/plugin.py:358: in __aexit__
>     await self._on_shutdown()
> ../../litestar/channels/plugin.py:341: in _on_shutdown
>     await self._sub_task
> ../../litestar/channels/plugin.py:310: in _sub_worker
>     async for channel, payload in self._backend.stream_events():
> ../../litestar/channels/backends/redis.py:105: in stream_events
>     message = await self._pub_sub.get_message(ignore_subscribe_messages=True, timeout=None)  # type: ignore[arg-type]
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/client.py:916: in get_message
>     response = await self.parse_response(block=(timeout is None), timeout=timeout)
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/client.py:795: in parse_response
>     response = await self._execute(conn, conn.read_response, timeout=read_timeout)
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/client.py:775: in _execute
>     return await conn.retry.call_with_retry(
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/retry.py:62: in call_with_retry
>     await fail(error)
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/client.py:764: in _disconnect_raise_connect
>     raise error
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/retry.py:59: in call_with_retry
>     return await do()
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/connection.py:836: in read_response
>     response = await self._parser.read_response(
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/connection.py:256: in read_response
>     response = await self._read_response(disable_decoding=disable_decoding)
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/connection.py:266: in _read_response
>     raw = await self._readline()
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <redis.asyncio.connection.PythonParser object at 0x7fd782220580>
> 
>     async def _readline(self) -> bytes:
>         """
>         read an unknown number of bytes up to the next '\r\n'
>         line separator, which is discarded.
>         """
>         found = self._buffer.find(b"\r\n", self._pos)
>         if found >= 0:
>             result = self._buffer[self._pos : found]
>         else:
>             tail = self._buffer[self._pos :]
>             data = await self._stream.readline()
>             if not data.endswith(b"\r\n"):
> >               raise ConnectionError(SERVER_CLOSED_CONNECTION_ERROR)
> E               redis.exceptions.ConnectionError: Connection closed by server.
> 
> /home/lotso/.cache/pypoetry/virtualenvs/litestar-34KzF20t-py3.11/lib/python3.11/site-packages/redis/asyncio/connection.py:343: ConnectionError
> ````
> 
> it seems to me that the issue is introduced by this commit (https://github.com/redis/redis-py/commit/e7306aae64b728fc6d894faed046960de76710c4 ) that changes the way redis "sees" if a connection is closed.
> 
> It now sees it as closed and the test fails at getting the subscibed data back, or there's something about this change I dont get ?
> 
> wdyt @Goldziher ?
> 

Best to involve @provinzkraut here, im less familiar with our listener logic than he is

**Comment by euri10** [comment]
the upstream offending lines imho if this helps:

https://github.com/redis/redis-py/commit/e7306aae64b728fc6d894faed046960de76710c4#diff-5924d2cbce9d62737f0024477d4a89f2314da1fd83da5b46dd02d85127d665cbL397-L402

**Comment by euri10** [comment]
from digging a little more it seems to me this breaks on the above lines because there is an empty byte sent at the very end `b""`  though I'm not sure where this could come from 

I cant see why we'd do that nor found an existing line that does it in Litestar codebase, however upstream SocketBuffer siginificantly changed between those two version so I suspect that may be the issue, again not sure about this, just giving thoughts maybe others will have a clearer idea :)

**Comment by provinzkraut** [comment]
> I cant see why we'd do that nor found an existing line that does it in Litestar codebase

I'm quite sure that we don't, since we don't even interact with redis on a low enough level to do this. I'm suspecting that, once again, an attempt to fix the original asyncio issue, another bug was introduced. If we can reliably reproduce this without the channels plugin, maybe we can file a bug report there?

**Comment by euri10** [comment]
actually my diagnosis was plain wrong and I have a fix... (note for myself to better read a stacktrace up to the top...)

in fact the bug is in the plugin, it's just a matter of closing the backend after the subscription task cancellation

File: tests/docker_service_fixtures.py [hunk 1]
```diff
@@ -53,7 +54,7 @@ def __init__(self, worker_id: str) -> None:
         self._base_command = [
             "docker",
             "compose",
-            "--file=tests/docker-compose.yml",
+            f"--file={Path(__file__).parent / 'docker-compose.yml'}",
```
Comment (euri10) [inline tests/docker_service_fixtures.py:57]: this is unrelated but I found it annoying to have in Pycharm to change my working directory to make this work, now you can run tests using that compose file from wherever working dir you want

**Review by provinzkraut — APPROVED** [review]

**Pull request merged**